### PR TITLE
Improved parameter handling on proxmox modules

### PR DESCRIPTION
**Pull request by russoz**
##### SUMMARY
While looking into the code for the purpose of reviewing #1763 , I noticed we could use ``required_if``, ``required_one_of`` and  ``required_together`` for some of the proxmox modules.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
plugins/modules/cloud/misc/proxmox.py
plugins/modules/cloud/misc/proxmox_kvm.py
plugins/modules/cloud/misc/proxmox_snap.py
plugins/modules/cloud/misc/proxmox_template.py
plugins/modules/cloud/misc/proxmox_user_info.py

##### ADDITIONAL INFORMATION

### CODE REVIEW

File: plugins/modules/cloud/misc/proxmox.py [hunk 1]
```diff
@@ -561,13 +563,7 @@ def main():
                 module.params[param] = value
 
     auth_args = {'user': api_user}
-    if not (api_token_id and api_token_secret):
-        # If password not set get it from PROXMOX_PASSWORD env
-        if not api_password:
-            try:
-                api_password = os.environ['PROXMOX_PASSWORD']
-            except KeyError as e:
-                module.fail_json(msg='You should set api_password param or use PROXMOX_PASSWORD environment variable')
```
Comment (russoz) [inline plugins/modules/cloud/misc/proxmox.py:570]: Documentation already points to use of environment variable.

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 2]
```diff
@@ -182,7 +182,7 @@ def main():
         argument_spec=dict(
             api_host=dict(required=True),
             api_user=dict(required=True),
-            api_password=dict(no_log=True),
+            api_password=dict(no_log=True, required=True, fallback=(env_fallback, ['PROXMOX_PASSWORD'])),
```
Comment (russoz) [inline plugins/modules/cloud/misc/proxmox_snap.py:185]: Does ansible test for `required` before or after trying the `fallback` mechanism? I am not sure on that one. If it tests _after_, then this code works as before.

Comment (felixfontein) [reply plugins/modules/cloud/misc/proxmox_snap.py:185]: Fallbacks are handled before doing anything else, so also before `required` is checked.

**Comment by felixfontein** [comment]
Dear maintainers, could you please check whether the new conditions are correct?

File: plugins/modules/cloud/misc/proxmox_user_info.py [hunk 3]
```diff
@@ -224,7 +224,7 @@ def main():
     module = AnsibleModule(
         argument_spec=module_args,
         required_one_of=[('api_password', 'api_token_id')],
-        required_together=[('api_token_id', 'api_token_secret')],
+        required_together=[('api_token_id', 'api_token_secret'), ('user', 'domain')],
```
Comment (Aversiste) [inline plugins/modules/cloud/misc/proxmox_user_info.py:227]: This part is not correct, `domain` can be used independently of `user` as shown in the module documentation. Perhaps a `required_if` or `required_by` could do the trick though?

**Review by Aversiste — COMMENTED** [review]
Looks like what my `module_utils/proxmox.py` is trying to achieve ;)

I agree with this PR as it is a step in the direction of better compatibility between the proxmox modules.

File: plugins/modules/cloud/misc/proxmox_user_info.py (hunk 3 above)
Comment (russoz) [reply plugins/modules/cloud/misc/proxmox_user_info.py:227]: Thanks for the review @Aversiste ! Will work on that tonight (GMT+13).

Comment (russoz) [reply plugins/modules/cloud/misc/proxmox_user_info.py]: Rolled back this one. Thanks again for the review.

**Comment by felixfontein** [comment]
@Aversiste can you take another look whether now everything is correct? Thanks :)

**Comment by Aversiste** [comment]
Looks good to me.

**Comment by felixfontein** [comment]
@russoz can you add a changelog fragment here as well? Since this changes code, it should have one (better declare it `bugfixes:`, though technically it isn't a bugfix).

**Comment by russoz** [comment]
steaming hot chglog frag coming out of the oven

File: changelogs/fragments/1765-proxmox-params.yml [hunk 4]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - proxmox - refactored some parameter validation code into use of ``env_fallback``, ``required_if``, ``required_together``, ``required_one_of`` (https://github.com/ansible-collections/community.general/pull/1765).
```
Comment (felixfontein) [inline changelogs/fragments/1765-proxmox-params.yml:2]: To make sure people know this applies not just to the module called `proxmox`, but also some of the others:
```suggestion
  - proxmox* modules - refactored some parameter validation code into use of ``env_fallback``, ``required_if``, ``required_together``, ``required_one_of`` (https://github.com/ansible-collections/community.general/pull/1765).
```

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@russoz thanks for improving this!
@Aversiste thanks for reviewing!

**Comment by felixfontein** [comment]
Manual backport in #1803 (stable-1 does not have proxmox_snap).

**Pull request merged**